In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from sklearn.metrics import mean_squared_error

In [ ]:
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

Mounted at /content/drive


In [ ]:
questionnaires = pd.read_excel('/content/drive/MyDrive/Colab Notebooks/проекты с ЯП/психпроект/triada_empathy_tolerance.xlsx')
questionnaires.head(5)

Unnamed: 0  Participant Private ID  Machiavellianism  Psychopathy  \
0           0                 6902136                11           18   
1           1                 6902485                20            7   
2           2                 6902731                15           17   
3           3                 6905533                11            5   
4           4                 6906816                 8            8   

   Narcissism  Perspective-taking scale  Fantasy scale  \
0          14                        25             23   
1          15                        14             31   
2          14                        25             30   
3          15                        22             35   
4          11                        25             34   

   Empathic concern scale  Personal distress scale  Tolerance to uncertainty  \
0                      22                       16                        62   
1                      20                       15                        64   
2                      11                       15                        72   
3                      27                       24                        50   
4                      34                       28                        62   

   Intolerance to uncertainty  Interpersonal intolerance to uncertainty  
0                          77                                        37  
1                          70                                        25  
2                          37                                        34  
3                          49                                        46  
4                          58                                        41

In [ ]:
del questionnaires['Unnamed: 0']

In [ ]:
questionnaires['Participant Private ID'].unique().shape[0]

62

In [ ]:
demography = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/проекты с ЯП/психпроект/a6tf.csv')
demography.head(5)

Participant Public ID  Participant Private ID  genderRU-quantised  \
0              w96zxzom               6905533.0                   2   
1              zu8icgg1               6908178.0                   2   
2              0pz0b5pc               7079747.0                   2   
3              f4mqa2t1               7079758.0                   2   
4              ooow2f93               7081909.0                   2   

   DoBRU-year  educationRU-quantised  
0        20.0                      3  
1        28.0                      5  
2        19.0                      5  
3        21.0                      4  
4        20.0                      3

In [ ]:
del demography['Participant Public ID']

In [ ]:
demography['Participant Private ID'].unique().shape[0]

62

In [ ]:
ultimatum = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/проекты с ЯП/психпроект/task_huav_data.csv')
ultimatum.head(5)

Participant Public ID  Participant Private ID  consent_fair_offer  \
0              7twy19yo               7079134.0                   6   
1              th29xh3g               7087745.0                   6   
2              gtz9hdm0               7098218.0                   6   
3              1vwa9zu0               7099724.0                   6   
4              b5crcyvi               7104366.0                   1   

   consent_conditionally_fair_offer  consent_unfair_offer  \
0                                 6                    15   
1                                 1                     0   
2                                 6                     5   
3                                 4                     7   
4                                 2                     7   

   rejection_fair_offer  rejection_conditionally_fair_offer  \
0                     0                                   0   
1                     0                                   5   
2                     0                                   0   
3                     0                                   2   
4                     4                                   4   

   rejection_unfair_offer  
0                       3  
1                      18  
2                      13  
3                      11  
4                      11

In [ ]:
del ultimatum['Participant Public ID']

In [ ]:
ultimatum['Participant Private ID'].unique().shape[0]

62

In [ ]:
data = pd.merge(pd.merge(questionnaires, demography, on='Participant Private ID'), ultimatum, on='Participant Private ID')
data.head(5)

Participant Private ID  Machiavellianism  Psychopathy  Narcissism  \
0                 6902136                11           18          14   
1                 6902485                20            7          15   
2                 6902731                15           17          14   
3                 6905533                11            5          15   
4                 6906816                 8            8          11   

   Perspective-taking scale  Fantasy scale  Empathic concern scale  \
0                        25             23                      22   
1                        14             31                      20   
2                        25             30                      11   
3                        22             35                      27   
4                        25             34                      34   

   Personal distress scale  Tolerance to uncertainty  \
0                       16                        62   
1                       15                        64   
2                       15                        72   
3                       24                        50   
4                       28                        62   

   Intolerance to uncertainty  Interpersonal intolerance to uncertainty  \
0                          77                                        37   
1                          70                                        25   
2                          37                                        34   
3                          49                                        46   
4                          58                                        41   

   genderRU-quantised  DoBRU-year  educationRU-quantised  consent_fair_offer  \
0                   1        33.0                      5                   6   
1                   2        18.0                      3                   6   
2                   2        24.0                      7                   5   
3                   2        20.0                      3                   6   
4                   2        20.0                      5                   6   

   consent_conditionally_fair_offer  consent_unfair_offer  \
0                                 1                     1   
1                                 6                     3   
2                                 6                    18   
3                                 6                     0   
4                                 6                    18   

   rejection_fair_offer  rejection_conditionally_fair_offer  \
0                     0                                   5   
1                     0                                   0   
2                     0                                   0   
3                     0                                   0   
4                     0                                   0   

   rejection_unfair_offer  
0                      17  
1                      15  
2                       0  
3                      18  
4                       0

In [ ]:
# 'DoBRU-year' consists null values
data.drop(data[data['DoBRU-year'].isnull() == True].index, inplace=True)

# Используем все предикторы для обучения модели

In [ ]:
data.columns

Index(['Participant Private ID', 'Machiavellianism', 'Psychopathy',
       'Narcissism', 'Perspective-taking scale', 'Fantasy scale',
       'Empathic concern scale', 'Personal distress scale',
       'Tolerance to uncertainty', 'Intolerance to uncertainty',
       'Interpersonal intolerance to uncertainty', 'genderRU-quantised',
       'DoBRU-year', 'educationRU-quantised', 'consent_fair_offer',
       'consent_conditionally_fair_offer', 'consent_unfair_offer',
       'rejection_fair_offer', 'rejection_conditionally_fair_offer',
       'rejection_unfair_offer'],
      dtype='object')

In [ ]:
X = data[['Participant Private ID', 'Machiavellianism', 'Psychopathy',
       'Narcissism', 'Perspective-taking scale', 'Fantasy scale',
       'Empathic concern scale', 'Personal distress scale',
       'Tolerance to uncertainty', 'Intolerance to uncertainty',
       'Interpersonal intolerance to uncertainty', 'genderRU-quantised',
       'DoBRU-year', 'educationRU-quantised']]

y = data[['consent_fair_offer',
       'consent_conditionally_fair_offer', 'consent_unfair_offer',
       'rejection_fair_offer', 'rejection_conditionally_fair_offer',
       'rejection_unfair_offer']]
#y_targets = np.array_split(y.to_numpy(), y.shape[0])

In [ ]:
X[X['DoBRU-year'].isnull() == True]

Empty DataFrame
Columns: [Participant Private ID, Machiavellianism, Psychopathy, Narcissism, Perspective-taking scale, Fantasy scale, Empathic concern scale, Personal distress scale, Tolerance to uncertainty, Intolerance to uncertainty, Interpersonal intolerance to uncertainty, genderRU-quantised, DoBRU-year, educationRU-quantised]
Index: []

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42)

rfr = RandomForestRegressor(max_depth=2, random_state=0)
rfr.fit(X_train, y_train)

RandomForestRegressor(max_depth=2, random_state=0)

In [ ]:
y_pred = rfr.predict(X_test).round()

In [ ]:
y_pred

array([[ 6.,  5., 11.,  0.,  1.,  7.],
       [ 6.,  5., 10.,  0.,  1.,  8.],
       [ 6.,  5.,  5.,  0.,  1., 13.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  5.,  0.,  1., 13.],
       [ 6.,  4.,  6.,  0.,  1., 12.],
       [ 6.,  4.,  4.,  0.,  2., 14.],
       [ 6.,  5.,  5.,  0.,  1., 13.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  6.,  0.,  1., 12.],
       [ 6.,  5.,  9.,  0.,  1.,  9.],
       [ 6.,  4.,  5.,  0.,  2., 13.],
       [ 6.,  4.,  4.,  0.,  1., 14.],
       [ 6.,  5.,  7.,  0.,  1., 11.],
       [ 6.,  5., 10.,  0.,  1.,  8.],
       [ 6.,  4.,  6.,  0.,  1., 12.]])

In [ ]:
mean_squared_error(y_test, y_pred)

13.491228070175438

In [ ]:
rfr2 = RandomForestRegressor(n_estimators=80, max_depth=2, random_state=0)
rfr2.fit(X_train, y_train)

y_pred2 = rfr2.predict(X_test).round()
mean_squared_error(y_test, y_pred2)

13.0

In [ ]:
rfr3 = RandomForestRegressor(n_estimators=80, max_depth=2, max_leaf_nodes=3, random_state=0)
rfr3.fit(X_train, y_train)

y_pred3 = rfr3.predict(X_test).round()
mean_squared_error(y_test, y_pred3)

12.903508771929827

In [ ]:
rfr4 = RandomForestRegressor(n_estimators=80, max_depth=2, max_leaf_nodes=2, random_state=0)
rfr4.fit(X_train, y_train)

y_pred4 = rfr4.predict(X_test).round()
mean_squared_error(y_test, y_pred4)

12.043859649122808

In [ ]:
rfr4_predictions = rfr4.predict(X).round()

In [ ]:
rfr4_columns_names = ['rfr_consent_fair_offer',
       'rfr_consent_conditionally_fair_offer', 'rfr_consent_unfair_offer',
       'rfr_rejection_fair_offer', 'rfr_rejection_conditionally_fair_offer',
       'rfr_rejection_unfair_offer']

data_rfr = pd.concat([data, pd.DataFrame(rfr4_predictions, columns=rfr4_columns_names)], axis=1)
data_rfr.head(5)

Participant Private ID  Machiavellianism  Psychopathy  Narcissism  \
0               6902136.0              11.0         18.0        14.0   
1               6902485.0              20.0          7.0        15.0   
2               6902731.0              15.0         17.0        14.0   
3               6905533.0              11.0          5.0        15.0   
4               6906816.0               8.0          8.0        11.0   

   Perspective-taking scale  Fantasy scale  Empathic concern scale  \
0                      25.0           23.0                    22.0   
1                      14.0           31.0                    20.0   
2                      25.0           30.0                    11.0   
3                      22.0           35.0                    27.0   
4                      25.0           34.0                    34.0   

   Personal distress scale  Tolerance to uncertainty  \
0                     16.0                      62.0   
1                     15.0                      64.0   
2                     15.0                      72.0   
3                     24.0                      50.0   
4                     28.0                      62.0   

   Intolerance to uncertainty  ...  consent_unfair_offer  \
0                        77.0  ...                   1.0   
1                        70.0  ...                   3.0   
2                        37.0  ...                  18.0   
3                        49.0  ...                   0.0   
4                        58.0  ...                  18.0   

   rejection_fair_offer  rejection_conditionally_fair_offer  \
0                   0.0                                 5.0   
1                   0.0                                 0.0   
2                   0.0                                 0.0   
3                   0.0                                 0.0   
4                   0.0                                 0.0   

   rejection_unfair_offer  rfr_consent_fair_offer  \
0                    17.0                     6.0   
1                    15.0                     6.0   
2                     0.0                     6.0   
3                    18.0                     6.0   
4                     0.0                     6.0   

   rfr_consent_conditionally_fair_offer  rfr_consent_unfair_offer  \
0                                   5.0                       9.0   
1                                   5.0                       8.0   
2                                   5.0                       9.0   
3                                   5.0                       8.0   
4                                   5.0                       8.0   

   rfr_rejection_fair_offer  rfr_rejection_conditionally_fair_offer  \
0                       0.0                                     1.0   
1                       0.0                                     1.0   
2                       0.0                                     1.0   
3                       0.0                                     1.0   
4                       0.0                                     1.0   

   rfr_rejection_unfair_offer  
0                         9.0  
1                         9.0  
2                         9.0  
3                        10.0  
4                        10.0  

[5 rows x 26 columns]

In [ ]:
data_rfr.drop(data_rfr.tail(1).index,inplace=True)

In [ ]:
y_sqrt = np.sqrt(y)

In [ ]:
X_train_sqrt, X_test_sqrt, y_train_sqrt, y_test_sqrt = train_test_split(
    X, y_sqrt, test_size=0.3, random_state=42)

rfr_sqrt = RandomForestRegressor(max_depth=2, random_state=0)
rfr_sqrt.fit(X_train_sqrt, y_train_sqrt)

y_pred_sqrt = rfr_sqrt.predict(X_test_sqrt).round()
mean_squared_error(y_test_sqrt, y_pred_sqrt)

0.9236967719596679

In [ ]:
rfr_sqrt2 = RandomForestRegressor(n_estimators=80, max_depth=2, max_leaf_nodes=2, random_state=0)
rfr_sqrt2.fit(X_train_sqrt, y_train_sqrt)

y_pred_sqrt2 = rfr_sqrt2.predict(X_test_sqrt).round()
mean_squared_error(y_test_sqrt, y_pred_sqrt2)

0.8225044253839742

In [ ]:
rfr_sqrt_predictions = rfr_sqrt2.predict(X).round()**2

In [ ]:
rfr_sqrt2_columns_names = ['rfr_sqrt_consent_fair_offer',
       'rfr_sqrt_consent_conditionally_fair_offer', 'rfr_sqrt_consent_unfair_offer',
       'rfr_sqrt_rejection_fair_offer', 'rfr_sqrt_rejection_conditionally_fair_offer',
       'rfr_sqrt_rejection_unfair_offer']

data_rfr_sqrt = pd.concat([data_rfr, pd.DataFrame(rfr_sqrt_predictions, columns=rfr_sqrt2_columns_names)], axis=1)
data_rfr_sqrt.head(5)

Participant Private ID  Machiavellianism  Psychopathy  Narcissism  \
0               6902136.0              11.0         18.0        14.0   
1               6902485.0              20.0          7.0        15.0   
2               6902731.0              15.0         17.0        14.0   
3               6905533.0              11.0          5.0        15.0   
4               6906816.0               8.0          8.0        11.0   

   Perspective-taking scale  Fantasy scale  Empathic concern scale  \
0                      25.0           23.0                    22.0   
1                      14.0           31.0                    20.0   
2                      25.0           30.0                    11.0   
3                      22.0           35.0                    27.0   
4                      25.0           34.0                    34.0   

   Personal distress scale  Tolerance to uncertainty  \
0                     16.0                      62.0   
1                     15.0                      64.0   
2                     15.0                      72.0   
3                     24.0                      50.0   
4                     28.0                      62.0   

   Intolerance to uncertainty  ...  rfr_consent_unfair_offer  \
0                        77.0  ...                       9.0   
1                        70.0  ...                       8.0   
2                        37.0  ...                       9.0   
3                        49.0  ...                       8.0   
4                        58.0  ...                       8.0   

   rfr_rejection_fair_offer  rfr_rejection_conditionally_fair_offer  \
0                       0.0                                     1.0   
1                       0.0                                     1.0   
2                       0.0                                     1.0   
3                       0.0                                     1.0   
4                       0.0                                     1.0   

   rfr_rejection_unfair_offer  rfr_sqrt_consent_fair_offer  \
0                         9.0                          4.0   
1                         9.0                          4.0   
2                         9.0                          4.0   
3                        10.0                          4.0   
4                        10.0                          4.0   

   rfr_sqrt_consent_conditionally_fair_offer  rfr_sqrt_consent_unfair_offer  \
0                                        4.0                            9.0   
1                                        4.0                            9.0   
2                                        4.0                            9.0   
3                                        4.0                            9.0   
4                                        4.0                            4.0   

   rfr_sqrt_rejection_fair_offer  rfr_sqrt_rejection_conditionally_fair_offer  \
0                            0.0                                          1.0   
1                            0.0                                          1.0   
2                            0.0                                          1.0   
3                            0.0                                          1.0   
4                            0.0                                          1.0   

   rfr_sqrt_rejection_unfair_offer  
0                              9.0  
1                              9.0  
2                              9.0  
3                              9.0  
4                              9.0  

[5 rows x 32 columns]

In [ ]:
data_rfr_sqrt.to_excel('Random Forest Ultimatum.xlsx')

# Используем только некоторые предикторы для обучения модели

In [ ]:
# уберем признаки "Tolerance to uncertainty" и "Intolerance to uncertainty"
X = data[['Participant Private ID', 'Machiavellianism', 'Psychopathy',
       'Narcissism', 'Perspective-taking scale', 'Fantasy scale',
       'Empathic concern scale', 'Personal distress scale',
       'Interpersonal intolerance to uncertainty', 'genderRU-quantised',
       'DoBRU-year', 'educationRU-quantised']]

y = data[['consent_fair_offer',
       'consent_conditionally_fair_offer', 'consent_unfair_offer',
       'rejection_fair_offer', 'rejection_conditionally_fair_offer',
       'rejection_unfair_offer']]

In [ ]:
X[X['DoBRU-year'].isnull() == True]

Empty DataFrame
Columns: [Participant Private ID, Machiavellianism, Psychopathy, Narcissism, Perspective-taking scale, Fantasy scale, Empathic concern scale, Personal distress scale, Interpersonal intolerance to uncertainty, genderRU-quantised, DoBRU-year, educationRU-quantised]
Index: []

In [ ]:
# далее повторим все предыдущие шаги
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=42)

rfr_new = RandomForestRegressor(n_estimators=80, max_depth=2, max_leaf_nodes=2, random_state=0)
rfr_new.fit(X_train, y_train)

y_pred_new = rfr_new.predict(X_test).round()
mean_squared_error(y_test, y_pred_new)

11.842105263157896

In [ ]:
rfr_new2 = RandomForestRegressor(max_depth=2, max_leaf_nodes=2, random_state=0)
rfr_new2.fit(X_train, y_train)

y_pred_new2 = rfr_new2.predict(X_test).round()
mean_squared_error(y_test, y_pred_new2) # аналогичный результат при n_estimators=100

11.614035087719296

In [ ]:
rfr_new3 = RandomForestRegressor(n_estimators=137, max_depth=2, max_leaf_nodes=2, random_state=0)
rfr_new3.fit(X_train, y_train)

y_pred_new3 = rfr_new3.predict(X_test).round()
mean_squared_error(y_test, y_pred_new3) # аналогичный результат при n_estimators=136

11.491228070175438

Таким образом, убрав 2 признака, удалось **уменьшить ошибку модели** с 12.043859649122808 до 11.491228070175438, или **на 0.55263157894737**.

In [ ]:
rfr_new_predictions = rfr_new3.predict(X).round()

rfr_new_columns_names = ['rfr_consent_fair_offer',
       'rfr_consent_conditionally_fair_offer', 'rfr_consent_unfair_offer',
       'rfr_rejection_fair_offer', 'rfr_rejection_conditionally_fair_offer',
       'rfr_rejection_unfair_offer']

data_rfr_new = pd.concat([data, pd.DataFrame(rfr_new_predictions, columns=rfr_new_columns_names)], axis=1)
data_rfr_new.head(5)

Participant Private ID  Machiavellianism  Psychopathy  Narcissism  \
0               6902136.0              11.0         18.0        14.0   
1               6902485.0              20.0          7.0        15.0   
2               6902731.0              15.0         17.0        14.0   
3               6905533.0              11.0          5.0        15.0   
4               6906816.0               8.0          8.0        11.0   

   Perspective-taking scale  Fantasy scale  Empathic concern scale  \
0                      25.0           23.0                    22.0   
1                      14.0           31.0                    20.0   
2                      25.0           30.0                    11.0   
3                      22.0           35.0                    27.0   
4                      25.0           34.0                    34.0   

   Personal distress scale  Tolerance to uncertainty  \
0                     16.0                      62.0   
1                     15.0                      64.0   
2                     15.0                      72.0   
3                     24.0                      50.0   
4                     28.0                      62.0   

   Intolerance to uncertainty  ...  consent_unfair_offer  \
0                        77.0  ...                   1.0   
1                        70.0  ...                   3.0   
2                        37.0  ...                  18.0   
3                        49.0  ...                   0.0   
4                        58.0  ...                  18.0   

   rejection_fair_offer  rejection_conditionally_fair_offer  \
0                   0.0                                 5.0   
1                   0.0                                 0.0   
2                   0.0                                 0.0   
3                   0.0                                 0.0   
4                   0.0                                 0.0   

   rejection_unfair_offer  rfr_consent_fair_offer  \
0                    17.0                     6.0   
1                    15.0                     6.0   
2                     0.0                     6.0   
3                    18.0                     6.0   
4                     0.0                     6.0   

   rfr_consent_conditionally_fair_offer  rfr_consent_unfair_offer  \
0                                   5.0                       9.0   
1                                   5.0                       9.0   
2                                   5.0                      10.0   
3                                   5.0                       8.0   
4                                   5.0                       8.0   

   rfr_rejection_fair_offer  rfr_rejection_conditionally_fair_offer  \
0                       0.0                                     1.0   
1                       0.0                                     1.0   
2                       0.0                                     1.0   
3                       0.0                                     1.0   
4                       0.0                                     1.0   

   rfr_rejection_unfair_offer  
0                         9.0  
1                         9.0  
2                         8.0  
3                        10.0  
4                        10.0  

[5 rows x 26 columns]

In [ ]:
data_rfr_new.drop(data_rfr_new.tail(2).index, inplace=True)

In [ ]:
data_rfr_new.to_excel('Random Forest Ultimatum without "Tolerance to uncertainty" & "Intolerance to uncertainty" features.xlsx')